caffeine@patapon.info/mprisMediaPlayer2.js

```diff
@@ -0,0 +1,322 @@
+            return;
+        }
+        this._lastEmittedPlayStatus = this.isPlaying;
+        this.emit('isPlaying', this.isPlaying);
+    }
+
+    /**
+     * @param {string} dbusName Name in dbus
+     */
+    _addPlayer(dbusName) {
+        if (this._activePlayers.has(dbusName)) {
+            return;
+        }
+
+        const dbusPlayerProxy = new this._DBusPlayerProxy(
+            Gio.DBus.session,
+            dbusName,
+            '/org/mpris/MediaPlayer2',
+            (_player) => this._onPlayerChange()
+        );
+
+        const handlerId = dbusPlayerProxy.connect(
+            'g-properties-changed',
+            (_player) => this._onPlayerChange()
+        );
+
+        this._activePlayers.set(dbusName, {
+            handlerId,
+            playerProxy: dbusPlayerProxy
+        });
+    }
+
+    /**
+     * @param {string} dbusName Name in dbus
+     */
+    _removePlayer(dbusName) {
+        const player = this._activePlayers.get(dbusName);
+        if (!player) {
+            return;
+        }
+        player.playerProxy.disconnect(player.handlerId);
+        this._activePlayers.delete(dbusName);
+    }
+
+    _onPlayerChange() {
+        let isPlaying = false;
+        for (const player of this._activePlayers.values()) {
+            if (player.playerProxy.PlaybackStatus === 'Playing') {
+                isPlaying = true;
+            }
+        }
+        this._isPlaying = isPlaying;
+        this._emitPlayStatus();
+    }
+
+    _onNameOwnerChanged(_proxy, _sender, [name, oldOwner, newOwner]) {
+        if (!name.startsWith(this._mprisPrefix)) {
+            return;
+        }
+        if (newOwner === "") {
+            this._removePlayer(name);
+        } else if (oldOwner === "") {
+            this._addPlayer(name);
+        }
+        this._onPlayerChange();
+    }
+
+    /**
+     * Get the dbus name list for mpris players
+     * @returns {string[]}
+     */
+    _getMPlayerApps() {
+        const [names] = this._dbusProxy.ListNamesSync();
+        const mprisPlayers = names.filter((dbusName) =>
+            dbusName.startsWith(this._mprisPrefix)
+        );
+
+        return mprisPlayers;
+    }
+
+    _onDestroy() {
+        this._dbusProxy.disconnectSignal(this._dbusHandlerId);
+        for (const dbusName of this._activePlayers.keys()) {
+            this._removePlayer(dbusName);
+        }
+        this._activePlayers.clear();
+
+        for (const connectId of this._connections.values()) {
+            this.disconnectIsPlaying(connectId);
+        }
+        this._connections.clear();
+    }
+
+    constructor() {
+        super();
+        // Declare some GObject.Object properties (non typescript has some difficulty with this weird import)
+        if (!this.emit) {
+            this.emit = super.emit;
+        }
+        if (!this.connect) {
+            this.connect = super.connect;
+        }
+        if (!this.disconnect) {
+            this.disconnect = super.disconnect;
+        }
```

Comment:
This seems odd to me, is this because the class isn't registered with `GObject.registerClass`?

Reply:
I removed this. This was for type checking failures. Serves no purpose anymore.